File: scripts/algorithms/B/Binary Tree Paths/Binary Tree Paths.java
// Runtime: 1 ms (Top 99.96%) | Memory: 42.40 MB (Top 35.43%)

class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> result = new ArrayList<>();
        dfs(root, new StringBuilder(), result);
        return result;
    }

    private void dfs(TreeNode node, StringBuilder path, List<String> result) {
        if (node == null) return;
        int len = path.length();
        if (len > 0) {
            path.append("->");
        }
        path.append(node.val);
        if (node.left == null && node.right == null) {
            result.add(path.toString());
        } else {
            dfs(node.left, path, result);
            dfs(node.right, path, result);
        }
        path.setLength(len); // Backtrack by resetting the StringBuilder
    }
}

File: scripts/algorithms/B/Binary Tree Paths/Binary Tree Paths.cpp
// Runtime: 3 ms (Top 89.27%) | Memory: 13.9 MB (Top 17.11%)
class Solution {
private:
    void dfs(TreeNode* root,string s,vector<string> &ans){
        if(root == NULL) return;

        s += (to_string(root -> val));

        if(root -> left != NULL || root -> right != NULL) s += "->";

        if(root -> left == NULL and root -> right == NULL) {
            ans.push_back(s);
        }

        dfs(root -> left,s,ans);
        dfs(root -> right,s,ans);
        s.pop_back();
    }
public:
    vector<string> binaryTreePaths(TreeNode* root) {

        vector<string> ans;
        string s;
        dfs(root,s,ans);
        return ans;

    }
};